chalice/deploy/planner.py

```diff
@@ -155,6 +216,13 @@ def plan_managediamrole(self, resource):
                         'policy_name': resource.role_name,
                         'policy_document': document},
                 resource=resource
+            ),
+            models.Pop(),
+            models.Push(resource.role_arn),
```

Comment:
So when is it recommended to the `Pop` and `Push` as opposed to a `StoreValue`?  The way I could have seen this code go is not do the whole Pop/Push workflow and just have the `APICall` and `RecordResourceValue` that provided the `resource.role_arn` directly. Seems like we need to do an unnecessary amount of hops to provide the `resource.role_arn` to the `RecordResourceValue` object.

Reply:
* `StoreValue` - If you're going to reference it later either in later instructions or as a param to an `APICall` instruction
* `Push` - Used in the next instruction.

It's a tradeoff between how simple your instructions are (easier executor to implement but more instructions to accomplish something) vs. more complex instructions with slightly more complicated executor.

The `RecordResourceValue` currently has two modes:

* It can record a value from a variable name (something that you've previously `StoreValue`'d)
* It can record a value that's at the top of the stack

What you'd be proposing here is a third mode:

* Record a value that's specified directly as an argument to this instruction.

It's doable, it's just a tradeoff of how complex we want to make the instruction.  The argument could go the other way as well.  "Why have a third mode when you can already accomplish this with the existing instructions."

Reply:
So looking at it again, I think my biggest issue is with direct exposure of the stack. Thinking through the interface from the perspective of someone new adding deployment support for a new resource, the biggest concern was the exposed stack interface and most of my suggestions (this third mode and the chaining suggestion) were related to how I would want to avoid it.

Specifically, my concerns with interacting with the stack were:
* If I pushed something onto the stack in an `execute()` call, what is the guarantee that the value pushed on will not affect a subsequent `execute()` call? And does that mean that I have to handle cleanup of what was pushed onto the stack? Right now this is protected by all of the `execute()` calls that we are doing starts with an `APICall()`, but there is no guarantee in the future. This concern was also furthered by the existence of a `Pop()` instruction that seemed to exist for the sole purpose of cleaning up values as no value is returned from the `Pop()` instruction.
* The stack interface opens the door for some poorly designed `execute()` workflows. Ideally, users of the interface should process the value for variable storage or resource recording immediately after the `Push()` or `APICall()`. But nothing is stopping them from making multiple  back to back `APICall()` or `Push()` and not immediately process the value for variable storage. Then at that point, it gets complicated because they need to keep track of the current state of all values in the stack. So like if they popped off a value, later down the road I would need to be aware that if they are storing is a variable or recording a resource value it will be of value that was underneath of what I had popped off.


I think a lot of these concerns would be satisfied if the stack was switched to more of a register, where:
* The register stores the value from the previous `APICall`. So if you want to store it or record it do so then as after the next `APICall` it will be a new value.
* On the call of every `execute()`, the register is cleared to start so there is no previous state.

That way it forces me to interact with it more in the way that is recommended and I do not have to worry about state carrying over.


Then as to the comment about the third mode of `RecordResourceValue`, could we just have different classes for these instructions? So to make sure the instruction class is not too complicated so we could break this out to a `RecordResourceFromValue`, `RecordResourceFromVariable`, and `RecordResourceFromStack` (or register or whatever comes out of this).

Reply:
> If I pushed something onto the stack in an execute() call, what is the guarantee that the value pushed on will not affect a subsequent execute() call? And does that mean that I have to handle cleanup of what was pushed onto the stack?

This isn't how the executor is intended to be used.  You build up an entire list of instructions and you ask the executor to execute it.  If we wanted to support multiple calls to `Executor.execute()` we could just add a `clear()` or a `reset()` state method.

Additionally, I don't see how your suggestion of register resolves this issue.  This concern is about leftover state not being cleaned up in subsequent `execute()` calls.  You have that whether you use a stack or registers.

> The stack interface opens the door for some poorly designed execute() workflows.

So does registers.  If you screw up the set of instructions to execute, you're going to have a problem.  Nothing is stopping multiple `APICall()s` from writing to the same register and thereby wiping out a variable that was previously written.

> So like if they popped off a value, later down the road I would need to be aware that if they are storing is a variable or recording a resource value it will be of value that was underneath of what I had popped off.

I don't understand what this means.  If you want to store a value, then don't pop something off the stack.  Who's "they" and "I" in this case?  No one's going to inject arbitrary instructions in the middle of the list of instructions needed to modify a resource.

> That way it forces me to interact with it more in the way that is recommended and I do not have to worry about state carrying over.

Not sure what this means either.  What is the "way that is recommended"?  State will carry over in either approach.  The way to solve that is with some method to clear out state.


I'm not opposed to removing the stack, but none of these arguments make a convincing case that using registers alleviate any of these problems.  In fact the typical tradeoff with a stack vs. register-based approach is that a stack is much easier to implement (and emit) at the cost of slightly more instructions.  I would like to know why *that* isn't the right tradeoff here.

Reply:
Talked about this offline. Given this rewrite is a work in-progress, we will go ahead with this approach because it is simpler to implement and there is future code that will utilize the interface more. We will look into refactoring this once more of the deployment logic for the other resources are fleshed out using the interface